06x-new-kafka-roller.md

```diff
@@ -0,0 +1,331 @@
+6. The roller restarts `controller-1` as it has no impact on the quorum health. When it has `SERVING` state, the roller repeats the quorum check and restarts `controller-2` and then `controller-0`. 
+7. It then considers restarting `combined-4`, so it performs quorum healthcheck and then availability check. Both check passes therefore `combined-4` is restarted. The same is repeated for `combined-5`.
+8. All controller and combined nodes have `SERVING` state and `numRestartAttempts` set to 1, therefore the roller checks the broker nodes.
+9. It sends a request to describe all the topic partitions, and the following list of topics is returned:
+```
+topic("topic-A"), Replicas(9, 10, 11), ISR(9, 10)
+topic("topic-B"), Replicas(6, 7, 8), ISR(6, 7, 8)
+topic("topic-C"), Replicas(10, 8, 6), ISR(10, 8, 6)
+topic("topic-D"), Replicas(7, 9, 11), ISR(7, 9, 11)
+topic("topic-E"), Replicas(6, 10, 11), ISR(6, 10, 11)
+```
+They are configured with `min.insync.replica` of 2.
+
+10. The roller batches the nodes that do not have any topic partition in common and the following batches are created: 
+- (11, 8) - `broker-11` and `broker-8` do not share any topic partitions.
+- (7) - `broker-7` and `broker-10` do not share any topic partitions, however topic-A is at min ISR, therefore 10 cannot be restarted and is removed from the batch.
+- (6) - `broker-6` and `broker-9` do not share any topic partitions, however topic-A is at min ISR, therefore 9 cannot be restarted and is removed from the batch.
+
+11. The roller picks the largest batch containing `broker-11` and `broker-8` and restarts them. It waits forthe nodes to have `SERVING` and then `LEADING_ALL_PREFERRED` state.
+12. It then restarts the batch containing only `broker-7` and restart it. It waits for it to have `SERVING` and then `LEADING_ALL_PREFERRED` state.
+13. It then restarts the batch containing only `broker-6` and restart it. It times out waiting for it to have `SERVING` state because it's still performing log recovery.
+14. The roller retries waiting for `broker-6` to have `SERVING` state for a number of times and results in the following context:
+```
+         nodeRef: broker-6/6
+         nodeRoles: broker
+         state: RECOVERING
+         lastTransition: 987456
+         reason:
+         numRestartAttempts: 1
+         numReconfigAttempts: 0
+         numRetries: 10
+```
+15. The `max_retries` of 10 is reached for `broker-6`, therefore the roller throws `UnrestartableNodesException` and the reconciliation fails. The operator logs the number of remaining segments and logs to recover.
+16. When the next reconciliation starts, all the nodes are observed and their contexts are updated. `broker-6` node has finished performing log recovery therefore have `SERVING` state. All nodes have no reason to restart. 
```

Comment:
Brokers 10 and 9 have not been restarted

Reply:
Good point! I have added them now.